File: Application_Source/main/java/com/example/go/myapplication03/MakeRSA.kt
import kotlin.random.Random
import java.math.BigInteger

class MakeRSA {
    fun generateRandomPrime(): Int {
        while (true) {
            val num = Random.nextInt(100, 1000)
            if (num < 2) {
                return -1
            }

            for (i in 2..(Math.sqrt(num.toDouble())).toInt() + 1) {
                if (num % i == 0) {
                    return -1
                }
            }
            return num
        }
    }

    fun generateKeypair(p: Int, q: Int): List<List<Int>> {
        val n = p * q
        val phi = (p - 1) * (q - 1)
        var e = Random.nextInt(1, phi)

        fun gcd(a: Int, b: Int): Int {
            return if (b == 0) a else gcd(b, a % b)
        }

        var g = gcd(e, phi)

        while (g != 1) {
            e = Random.nextInt(1, phi)
            g = gcd(e, phi)
        }

        var d = 0
        var x1 = 0
        var x2 = 1
        var y1 = 1
        var tempPhi = phi
        val e1 = e

        while (e > 0) {
            val temp1 = tempPhi / e
            val temp2 = tempPhi - temp1 * e
            tempPhi = e
            e = temp2
            val x = x2 - temp1 * x1
            val y = d - temp1 * y1
            x2 = x1
            x1 = x
            d = y1
            y1 = y
        }

        if (tempPhi == 1) {
            d += phi
        }

        return listOf(listOf(e1, n), listOf(d, n))
    }

    fun encrypt(publicKey: List<Int>, plaintext: String): List<Int> {
        val key = publicKey[0]
        val n = publicKey[1]
        val cipher = plaintext.map { BigInteger.valueOf(it.toLong()).modPow(BigInteger.valueOf(key.toLong()), BigInteger.valueOf(n.toLong())).toInt() }
        return cipher
    }

    fun decrypt(privateKey: List<Int>, ciphertext: List<Int>): String {
        val key = privateKey[0]
        val n = privateKey[1]
        val plain = ciphertext.map { BigInteger.valueOf(it.toLong()).modPow(BigInteger.valueOf(key.toLong()), BigInteger.valueOf(n.toLong())).toInt().toChar() }
        return plain.joinToString("")
    }
}

File: blockchain/makeRSA.py
import random

class makeRSA:
    def generate_random_prime():
        while True:
            num = random.randint(100, 1000)
            if num < 2:
                return False

            for i in range(2, int(num ** 0.5) + 1):
                if num % i == 0:
                    return False
            return num

    def generate_keypair(p, q):
        n = p * q
        phi = (p - 1) * (q - 1)
        e = random.randrange(1, phi)
        def gcd(a, b):
            if b == 0:
                return a

            else:
                return gcd(b, a % b)
        g = gcd(e, phi)

        while g != 1:
            e = random.randrange(1, phi)
            g = gcd(e, phi)
        d = 0
        x1 = 0
        x2 = 1
        y1 = 1
        temp_phi = phi
        e1 = e

        while e > 0:
            temp1 = temp_phi // e
            temp2 = temp_phi - temp1 * e
            temp_phi = e
            e = temp2
            x = x2 - temp1 * x1
            y = d - temp1 * y1
            x2 = x1
            x1 = x
            d = y1
            y1 = y

        if temp_phi == 1:
            d = d + phi

        return ((e1, n), (d, n))

    def encrypt(public_key, plaintext):
        key, n = public_key
        cipher = [(ord(char) ** key) % n for char in plaintext]

        return cipher

    def decrypt(private_key, ciphertext):
        key, n = private_key
        plain = [chr((char ** key) % n) for char in ciphertext]

        return ''.join(plain)

